### DISCUSSION TITLE
Output image colors changed after ImageMagick upgrade

### DISCUSSION THREAD

**Discussion by hatanlinh** [ImageMagick/ImageMagick#2920] [Help] [answered]
We're using ImageMagick as an extension to handle image files in our project.

Normally, we use this command to process JPEG files, which are extracted directly from PDF:
`convert -colorspace CMYK -quality 80 -set colorspace:auto-grayscale off -strip jpg:input jpg:output`

After upgraded from version 7.0.8-14 to 7.0.10-35, we're noticed that the color of the converted image has changed, resulting in a serious problem for us. As you can see in these attachments:

- This is the original:
![original](https://user-images.githubusercontent.com/22725761/100960853-b62ea500-3553-11eb-8b1b-3c0160f13836.jpg)

- This is the output of version 7.0.8-14:
![converted_before](https://user-images.githubusercontent.com/22725761/100960914-d9595480-3553-11eb-94ad-2f9c326b3f1a.jpg)

- This is the output of version 7.0.10-35:
![converted_after](https://user-images.githubusercontent.com/22725761/100960948-e7a77080-3553-11eb-9d8c-2f1403b35944.jpg)

We analyzed metadata from both of these and found out that the original has `APP14 Flags 0` set to `Encoded with Blend=1 downsampling` while both the output had this flag empty. Everything else is looking identical.
I personally spent several hours debug the code from two version trying to find any differences, but I couldn't find any.

Any information/advice for this problem would be a great news for us.
Thank you for your consideration!

**Comment by fmw42** [comment]
Sorry, I am unclear about your problem. Is the input to your command the PDF or a JPG extracted from the PDF? Does the PDF contain a raster image or is it totally vector? If the former, perhaps you should use pdfimages to extract the JPG from the PDF before using any ImageMagick processing.

**Reply by hatanlinh** [reply]
Hi @fmw42 , sorry for the unclear explanation, the input is raster JPG file extracted from PDF.

**Reply by fmw42** [reply]
How did you extract it from the PDF?  What tool? In your command

convert -colorspace CMYK -quality 80 -set colorspace:auto-grayscale off -strip jpg:input jpg:output

since the input is raster, the input image should come right after convert

convert jpg:input -colorspace CMYK -quality 80 -set colorspace:auto-grayscale off -strip jpg:output

Is the jpg input sRGB or CMYK?  Does it have a profile? If it does, it should be converted to CMYK with a profile. You will get better color conversion that way.  Why are you using -set colorspace:auto-grayscale off and -colorspace CMYK. If the image is gray, -colorspace CMYK will not be appropriate.

Your input is already CMYK, so -colorspace CMYK should not be added

Both of these produce similar results on IM 6.9.11.45 and IM 7.0.10.45 using libjpg 90

Input:

![img](https://user-images.githubusercontent.com/13458218/101117947-f9613600-359c-11eb-9cc6-ee4f0eb239eb.jpg)


`convert jpg:img.jpg -colorspace CMYK -quality 80 -set colorspace:auto-grayscale off -strip jpg:result1.jpg`

![result1](https://user-images.githubusercontent.com/13458218/101118231-82786d00-359d-11eb-93e1-fd65acab5ad7.jpg)


`convert jpg:img.jpg -quality 80 -strip jpg:result2.jpg`

![result2](https://user-images.githubusercontent.com/13458218/101118236-85735d80-359d-11eb-9c0c-4c95722501ec.jpg)



`compare -metric rmse img.jpg result1.jpg null:`
1724.64 (0.0263164)

`compare -metric rmse img.jpg result2.jpg null:`
1724.64 (0.0263164)


Perhaps check your version of libjpg

`magick -list format`

Look at the end of the line for JPG

JPG* JPEG      rw-   Joint Photographic Experts Group JFIF format **(libjpeg 90)**

**Reply by Jmuccigr** [reply]
> Hi @fmw42 , sorry for the unclear explanation, the input is raster JPG file extracted from PDF.

I would second @fmw42's advice to use pdfimages to extract raster images from a PDF, if you are not already. I do a lot of this. Specifically you want:

`pdfimages -f <firstpage #> -l <lastpage #> -all input.PDF /path/to/output/prefix`

This will extract the images from pages f through l and save them as prefix-000.jpg and so on. It will maintain the original format if it can (that's what the "-all" does), working with jpg, png, tiff, and other types. It's quite fast, certainly much faster than using IM to recode the images.

I can't help with the specific problem you're having, but will note too that you should be using `magick`, not `convert` with version 7.

**Reply by hatanlinh** [reply]
We're using our own PDF parser to extract the images directly from `/DCTDecode` filter of PDF file. I've verified and could be sure that our extracted images are identical to which extracted by `pdfimages`.

About the command line options passed to ImageMagick, we check the flags `/DeviceCMYK` in the PDF image object, if it is set, we automatically add `-colorspace=CMYK` and `-set colorspace:auto-grayscale off` to the command line to convert extracted JPGs. Thank you for pointing out this incorrect usage.

We're using `libjpeg-turbo 2.0.0` (which doesn't support libjpeg 90 as stated in their website iirc), this is a legacy choice and I'm not sure the reason behind this except performance. Is it not OK using this, what are your recommendation?

Thank you for your help @fmw42 and @Jmuccigr .

**Reply by fmw42** [reply]
Others use libjpeg-turbo, so that should be OK. But is that the most current version? As was pointed out, for IM 7, you should be using magick not convert and command order is more strict in IM 7. So read the input first. What happens if you simply try

`magick jpg:img.jpg -quality 80 -strip jpg:result2.jpg`

or

`magick jpg:img.jpg  jpg:result2.jpg`

or try adding a CMYK-type profile

`magick jpg:img.jpg -profile USWebCoatedSWOP.icc jpg:result2.jpg`

**Reply by hatanlinh** [reply]
hi @fmw42 ,
Thanks to your suggestion, I've checked our project code flow and found that this JPG image is actually associated with an ICC profile.
Normally, we would pass the ICC profile data to an ICC reader (ICCProLib) to parse and get the colorspace field in the header, then set the `-colorspace` option to IM call accordingly. Other information in the the ICC profile are discarded, as it is unneeded for our use case, previously.

I've tried to write ICC profile data into an ICC file to run your suggested command `magick jpg:img.jpg -profile USWebCoatedSWOP.icc jpg:result2.jpg`

Here is the file: 
[ycmk_image.zip](https://github.com/ImageMagick/ImageMagick/files/5651667/ycmk_image.zip)
But this file looks... strange. Over 500KB in file size. It's also could not be opened by ICC Profile Inspector. Inspect with Exiftool gives only the following information:
```
ExifTool Version Number         : 12.12
File Name                       : ycmk_image.icc
Directory                       : D:/OPSWAT/workspace/CDS-627/Ne
File Size                       : 546 KiB
File Modification Date/Time     : 2020:12:07 14:00:39+07:00
File Access Date/Time           : 2020:12:07 16:14:49+07:00
File Creation Date/Time         : 2020:12:07 14:00:39+07:00
File Permissions                : rw-rw-rw-
File Type                       : ICC
File Type Extension             : icc
MIME Type                       : application/vnd.iccprofile
Profile CMM Type                : Adobe Systems Inc.
Profile Version                 : 2.1.0
Profile Class                   : Output Device Profile
Color Space Data                : CMYK
Profile Connection Space        : Lab
Profile Date Time               : 2000:07:26 05:41:53
Profile File Signature          : acsp
Primary Platform                : Apple Computer Inc.
CMM Flags                       : Not Embedded, Independent
Device Manufacturer             : Adobe Systems Inc.
Device Model                    :
Device Attributes               : Reflective, Glossy, Positive, Color
Rendering Intent                : Perceptual
Connection Space Illuminant     : 0.9642 1 0.82491
Profile Creator                 : Adobe Systems Inc.
Profile ID                      : 0
-- press ENTER --
```
I've tried convert the JPG file with this ICC profile, but the output is broken.

**Reply by snibgo** [reply]
iccDumpProfile says the file ycmk_image.icc is not a valid ICC profile file.

```
f:\web\im>iccxml\iccDumpProfile.exe ycmk_image.icc

Profile:          'ycmk_image.icc'
Profile ID:       Profile ID not calculated.
Size:             557168(0x88070) bytes

Header
------
Attributes:       Reflective | Glossy
Cmm:              Adobe
Creation Date:    7/26/2000  05:41:53
Creator:          'ADBE' = 41444245
Data Color Space: CmykData
Flags             EmbeddedProfileFalse | UseAnywhere
PCS Color Space:  LabData
Platform:         Macintosh
Rendering Intent: Perceptual
Type:             OutputClass
Version:          2.10
Illuminant:       X=0.9642, Y=1.0000, Z=0.8249

Profile Tags
------------
                      Tag    ID      Offset         Size
                     ----  ------    ------         ----
Unknown '?des' = 0A646573  '?des'  1660944384   -67108864
Unknown 'tcpr' = 74637072  'tcpr'  1946157057   1879048192
Unknown '+wtp' = 2B777470  '+wtp'  1946157057   -1677721600
Unknown '?A2B' = 14413242  '?A2B'  805306369    -1342177118
Unknown '?A2B' = 06413242  '?A2B'  838860801    -1342177118
Unknown '?A2B' = 06413242  '?A2B'  822083747    -1207959390
Unknown '?B2A' = 06423241  '?B2A'  805306693    -1073741256
Unknown '?B2A' = B4423241  '?B2A'  822084478    1946157624
Unknown '?B2A' = B4423241  '?B2A'  838862263    671089208
Unknown '?gam' = B467616D  '?gam'  1946159087   -603979632
Unknown '?des' = 91646573  '?des'  1660944384          0
Unknown '?U.S' = 1A552E53  '?U.S'  773871461    1646281583
Unknown 'ated' = 61746564  'ated'  539513687    1330653472
```
Note the bad tag IDs, and stupid offsets and lengths within the file.

**Comment by dlemstra** [comment, marked as answer]
What happens when you upgrade to the most recent version?

**Reply by hatanlinh** [reply]
I didn't try that, due to the integration of IM into our project, upgrading will require some effort. But I will try and tell you the result if it's possible.

**Reply by dlemstra** [reply]
Just noticed that you wrote about the `APP14` marker. I think you will need to upgrade to a more recent version. I suspect that this commit fixed your issue: https://github.com/ImageMagick/ImageMagick/commit/93f4a4320a3c2526a5a56d3e3adeb836c18bdd88.

**Reply by hatanlinh** [reply]
Thank you so much for your suggestion, this commit really fixed the problem.

**Comment by snibgo** [comment]
As fmw42 says, the OP command...
> convert -colorspace CMYK -quality 80 -set colorspace:auto-grayscale off -strip jpg:input jpg:output

... is in the wrong order. This may explain why the behaviour changes between IM versions.

What is the objective of that command? It reads and decompresses a JPG, strips metadata, and re-compresses it. This doesn't seem very useful.

The input file is CMYK with no embedded ICC profile. This is bad news, as software that reads the file don't know what primaries to use for the colour channels. Any software that displays the image on the screen must make assumptions about the CMYK colorspace, and those assumptions are likely to be wrong. For RGB files with no profile, the default assumption is usually sRGB. For CMYK with no profile, there is no widely adopted default.

I don't know much about raster images in PDF files. In particular, I don't know if they ever contain embedded profiles.

**Comment by fmw42** [comment]
I do not understand why you try to get the information from the profile and then remove it and then try to use -colorspace.  This is not going to preserve your colors properly.

**Comment by hatanlinh** [comment]
I would like to give a big thanks to everyone in this thread for discussing the issue and pointing out flaws in our image processing flow.
We'll take your suggestionsto review and update our project properly.
